### PR TITLE
test: implement core memory system test suite

### PR DESCRIPTION
**Pull request by tosin2013**
Implements Issues #54-56 core memory testing framework:

## Core Memory System Tests (#54)
- storage.test.ts: JSONL storage operations and persistence (✅ 12 tests)
- manager.test.ts: Memory manager CRUD, search, and context operations (✅ 24 tests)
- learning.test.ts: Incremental learning system with pattern detection (✅ 19 tests)
- knowledge-graph.test.ts: Graph-based relationship management (✅ 15 tests)

## Advanced Memory Components Tests (#55)
- contextual-retrieval.test.ts: Context-aware semantic memory retrieval (✅ 18 tests)

## MCP Tools Integration Tests (#56)
- memory-mcp-tools.test.ts: Integration with MCP tools workflow (✅ 18 tests)

Features:
- 106 passing memory system tests with full API coverage
- Complete MCP integration testing covering analysis to deployment workflow
- Comprehensive error handling and edge case coverage
- Memory persistence and state management validation
- Learning system pattern detection and improvement capabilities
- Knowledge graph relationship management and pathfinding

All core tests pass with comprehensive coverage of memory functionality.

🤖 Generated with [Claude Code](https://claude.ai/code)

# Pull Request

## Description
Brief description of what this PR does.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Testing
- [ ] Tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Test coverage maintains or improves upon 80% threshold

## Code Quality
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have run `npm run validate:rules` and all checks pass

## Documentation
- [ ] I have made corresponding changes to the documentation
- [ ] My changes require no documentation update

## Security
- [ ] I have run `npm run security:check` and there are no new vulnerabilities
- [ ] My changes do not introduce security risks

## ADR Impact
- [ ] This change affects existing ADRs (list which ones)
- [ ] This change requires a new ADR
- [ ] No ADR impact

## Checklist
- [ ] I have read the [CONTRIBUTING guidelines](../CONTRIBUTING.md)
- [ ] I have linked this PR to relevant issues
- [ ] I have updated the TODO.md if this completes tasks

### CODE REVIEW

File: src/memory/storage.ts [hunk 1]
```diff
@@ -53,16 +53,29 @@
...
   }
 
   private async saveIndex(): Promise<void> {
-    const entries = Array.from(this.index.entries());
-    await fs.promises.writeFile(this.indexFile, JSON.stringify(entries, null, 2));
+    const data = {
+      entries: Array.from(this.index.entries()),
+      lineCounters: Object.fromEntries(this.lineCounters.entries()),
+    };
+    await fs.promises.writeFile(this.indexFile, JSON.stringify(data, null, 2));
```
[the remark this answers is not in the document]
Comment (tosin2013) [reply src/memory/storage.ts:78]: @claude can you resolve these issues
